### PR TITLE
Experimenting with recipes in CSV RDoc

### PR DESCRIPTION
**Pull request by BurdetteLamar**
New Recipes page is oriented to what the user wants to do (instead of what CSV can do).  User finds in Contents the relevant task, follows link to get example working code.

Fulsome discussion welcome.

### CODE REVIEW

File: doc/csv/recipes.rdoc [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+  CSV.parse(string, headers: true) # => #<CSV::Table mode:col_or_row row_count:4>
+
+Instance method CSV#each can read a source \String one row at a time:
+  CSV.new(string, headers: true).each do |row|
+    p row
+  end
+Ouput:
+  #<CSV::Row "Name":"foo" "Value":"0">
+  #<CSV::Row "Name":"bar" "Value":"1">
+  #<CSV::Row "Name":"baz" "Value":"2">#
```
Comment (kou) [inline doc/csv/recipes.rdoc:60]:
```suggestion
  #<CSV::Row "Name":"baz" "Value":"2">
```

File: doc/csv/recipes.rdoc [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+
+==== Parse from \File Without Headers
+
+Instance method CSV#read can read a file all at once:
+  string = "foo,0\nbar,1\nbaz,2\n"
+  CSV.new(string).read # => [["foo", "0"], ["bar", "1"], ["baz", "2"]]
```
Comment (kou) [inline doc/csv/recipes.rdoc:66]: `CSV.read(path)`?

File: doc/csv/recipes.rdoc [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+  - {Filter IO Stream to String Without Headers}[#label-Filter+IO+Stream+to+String+Without+Headers]
+  - {Filter IO Stream to String with Headers}[#label-Filter+IO+Stream+to+String+with+Headers]
+  - {Filter IO Stream to IO Stream Without Headers}[#label-Filter+IO+Stream+to+IO+Stream+Without+Headers]
+  - {Filter IO Stream to IO Stream with Headers}[#label-Filter+IO+Stream+to+IO+Stream+with+Headers]
+
+=== Parsing
+
+==== Parse from \String Without Headers
+
+\Class method CSV::parse can read a source \String all at once,
```
Comment (kou) [inline doc/csv/recipes.rdoc:32]: Could you use `#{CLASS_NAME}.#{CLASS_METHOD_NAME}` format?

```suggestion
\Class method CSV.parse can read a source \String all at once,
```

File: doc/csv/recipes.rdoc [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+  ["baz", "2"]
+
+==== Parse from \IO Stream with Headers
+
+\Class method CSV::parse can read an \IO stream all at once:
+  string = "Name,Value\nfoo,0\nbar,1\nbaz,2\n"
+  path = 't.csv'
+  File.write(path, string)
+  File.open(path, headers: true) do |file|
+    CSV.parse(file)
```
Comment (kou) [inline doc/csv/recipes.rdoc:124]:
```suggestion
  File.open(path) do |file|
    CSV.parse(file, headers: true)
```

We also need to update output.

File: doc/csv/recipes.rdoc [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+  end # => [["Name", "Value"], ["foo", "0"], ["bar", "1"], ["baz", "2"]]
+
+\Class method CSV::foreach can read one row at a time:
+  File.open(path, headers: true) do |file|
+    CSV.foreach(file) do |row|
```
Comment (kou) [inline doc/csv/recipes.rdoc:129]: ditto.

File: doc/csv/recipes.rdoc [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+  output_string # => "foo,0\nbar,1\nbaz,2\n"
+
+==== Generate to \String with Headers
+
+\Class method CSV::generate can generate to a \String.
+
+This example uses method CSV#<< to append the rows
+that are to be generated:
+  output_string = CSV.generate('', headers: ['Name', 'Value']) do |csv|
+    csv << csv.headers
```
Comment (kou) [inline doc/csv/recipes.rdoc:161]:
```suggestion
  output_string = CSV.generate('', headers: ['Name', 'Value'], write_headers: true) do |csv|
```

File: doc/csv/recipes.rdoc [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+This example uses method CSV#<< to append the rows
+that are to be generated:
+  output_string = CSV.generate do |csv|
+    csv << ['Foo', 0]
+    csv << ['Bar', 1]
+    csv << ['Baz', 2]
+  end
+  output_string # => "foo,0\nbar,1\nbaz,2\n"
+
+==== Generate to \String with Headers
```
Comment (kou) [inline doc/csv/recipes.rdoc:154]:
```suggestion
==== Generate to \String With Headers
```

File: doc/csv/recipes.rdoc [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+
+==== Generate to \File with Headers
+
+\Class method CSV::open can generate to a \File.
+
+This example uses method CSV#<< to append the rows
+that are to be generated:
+  path = 't.csv'
+  CSV.open(path, 'w', headers: ['Name', 'Value']) do |csv|
+    csv << csv.headers
```
Comment (kou) [inline doc/csv/recipes.rdoc:190]: ditto.

File: doc/csv/recipes.rdoc [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+  end
+  p File.read(path) # => "Foo,0\nBar,1\nBaz,2\n"
+
+==== Generate to \IO Stream with Headers
+
+\\Classs method CSV::new can generate \CSV data to an \IO stream:
+  path = 't.csv'
+  File.open(path, 'w') do |file|
+    csv = CSV.new(file, headers: ['Name', 'Value'])
+    csv << csv.headers
```
Comment (kou) [inline doc/csv/recipes.rdoc:215]:
```suggestion
    csv = CSV.new(file, headers: ['Name', 'Value'], write_headers: true)
```

File: doc/csv/recipes.rdoc [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+      row[1] *= 4
+    end
+  end
+  p File.read(path) # => "FOO,0000\nBAR,1111\nBAZ,2222\n"
+
+==== Filter \String to \IO Stream with Headers
+
+  in_string = "Name,Value\nfoo,0\nbar,1\nbaz,2\n"
+  path = 't.csv'
+  File.open(path, 'w', headers: true) do |out_io|
```
Comment (kou) [inline doc/csv/recipes.rdoc:263]:
```suggestion
  File.open(path, 'w') do |out_io|
```

File: doc/csv/recipes.rdoc [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+    CSV.filter(in_string, out_io, headers: true) do |row|
+      p row
```
Comment (kou) [inline doc/csv/recipes.rdoc:265]:
```suggestion
```

File: doc/csv/recipes.rdoc [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+  p File.read(path) # => "Name,Value\nFOO,0000\nBAR,1111\nBAZ,2222\n"
+
+==== Filter \IO Stream to \String Without Headers
+
+  in_string = "foo,0\nbar,1\nbaz,2\n"
+  path = 't.csv'
+  File.write(path, in_string)
+  out_string = ''
+  File.open(path) do |file|
+    CSV.filter(in_string, out_string) do |row|
```
Comment (kou) [inline doc/csv/recipes.rdoc:279]:
```suggestion
  File.open(path) do |in_io|
    CSV.filter(in_io, out_string) do |row|
```

File: doc/csv/recipes.rdoc [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+  out_string # => "FOO,0000\nBAR,1111\nBAZ,2222\n"
+
+==== Filter \IO Stream to \String with Headers
+
+  in_string = "Name,Value\nfoo,0\nbar,1\nbaz,2\n"
+  path = 't.csv'
+  File.write(path, in_string)
+  out_string = ''
+  File.open(path, headers: true) do |file|
+    CSV.filter(in_string, out_string, headers: true) do |row|
```
Comment (kou) [inline doc/csv/recipes.rdoc:293]:
```suggestion
  File.open(path) do |in_io|
    CSV.filter(in_io, out_string, headers: true) do |row|
```

**Comment by BurdetteLamar** [comment]
Thanks, @kou.  Did you like the idea of recipes?  If so, there are many more possibilities.

**Comment by kou** [comment]
Yes! I like this.

File: doc/csv/recipes.rdoc (hunk 1 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed.

File: doc/csv/recipes.rdoc (hunk 3 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed (throughout).

File: doc/csv/recipes.rdoc (hunk 9 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed.

File: doc/csv/recipes.rdoc (hunk 11 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed. (Oops, debugging statement.)

File: doc/csv/recipes.rdoc (hunk 13 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed.

File: doc/csv/recipes.rdoc [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+\Class method CSV.generate can generate to a \String.
+
+This example uses method CSV#<< to append the rows
+that are to be generated:
+  output_string = CSV.generate do |csv|
+    csv << ['Foo', 0]
+    csv << ['Bar', 1]
+    csv << ['Baz', 2]
+  end
+  output_string # => "foo,0\nbar,1\nbaz,2\n"
```
Comment (esparta) [inline doc/csv/recipes.rdoc:151]:
```suggestion
  output_string # => "Foo,0\nBar,1\nBaz,2\n"
```

File: doc/csv/recipes.rdoc [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+  - {Filter IO Stream to String with Headers}[#label-Filter+IO+Stream+to+String+with+Headers]
+  - {Filter IO Stream to IO Stream Without Headers}[#label-Filter+IO+Stream+to+IO+Stream+Without+Headers]
+  - {Filter IO Stream to IO Stream with Headers}[#label-Filter+IO+Stream+to+IO+Stream+with+Headers]
+
+=== Parsing
+
+==== Parse from \String Without Headers
+
+\Class method CSV.parse can read a source \String all at once,
+and so may have memory resource implications:
```
Comment (esparta) [inline doc/csv/recipes.rdoc:33]:
```suggestion
be aware this <procedure>  may have memory resource implications:
```
Not sure about adding procedure or something different.

**Review by esparta — COMMENTED** [review]
minimal typo & phrase suggestion

File: doc/csv/recipes.rdoc (hunk 14 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed, and thanks.

File: doc/csv/recipes.rdoc (hunk 15 above)
Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc:33]: I'm leaving this as is.  CSV.parse is the subject in both clauses, and so the sense, as I see it, is "CSV.parse may have  memory resource implications."

**Comment by BurdetteLamar** [comment]
@kou, this may be ready to merge.

File: doc/csv/recipes.rdoc (hunk 6 above)
Comment (kou) [reply doc/csv/recipes.rdoc:160]: Could you confirm this too?

Comment (BurdetteLamar) [reply doc/csv/recipes.rdoc]: Fixed in new push (plus, one more similar).

**Comment by kou** [comment]
Thanks!

**Pull request merged**